### PR TITLE
Support optimistic sync in ProtoArray

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Allows nodes to be added optimistically that are later marked valid or invalid with separate methods to find the strict and optimistic head.

Note that production code currently marks every block as `VALID` immediately (as will remain the case prior to the merge fork) so there's no change in actual behaviour.

## Fixed Issue(s)
#4599 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java [hunk 1]
```diff
@@ -171,18 +219,73 @@ public ProtoNode findHead(Bytes32 justifiedRoot, UInt64 justifiedEpoch, UInt64 f
...
+    final Optional<ProtoNode> maybeNode = getProtoNode(blockRoot);
+    if (maybeNode.isEmpty()) {
+      // Most likely just pruned prior to the validation result being received.
+      LOG.debug("Couldn't mark block {} valid because it was unknown", blockRoot);
+      return;
+    }
+    final ProtoNode node = maybeNode.get();
+    node.setValidationStatus(VALID);
+    Optional<Integer> parentIndex = node.getParentIndex();
+    while (parentIndex.isPresent()) {
```
Comment (ajsutton) [inline protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java:257]: Should stop iterating when the first valid node is found.

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java [hunk 2]
```diff
@@ -171,18 +219,73 @@ public ProtoNode findHead(Bytes32 justifiedRoot, UInt64 justifiedEpoch, UInt64 f
...
+    for (int i = index + 1; i < nodes.size(); i++) {
+      final ProtoNode possibleDescendant = nodes.get(i);
+      if (possibleDescendant.getParentIndex().isEmpty()) {
+        continue;
+      }
+      if (invalidParents.contains((int) possibleDescendant.getParentIndex().get())) {
+        possibleDescendant.setValidationStatus(INVALID);
+        invalidParents.add(i);
+      }
+    }
```
Comment (ajsutton) [inline protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java:295]: Extract as a separate method

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java [hunk 3]
```diff
@@ -171,18 +219,73 @@ public ProtoNode findHead(Bytes32 justifiedRoot, UInt64 justifiedEpoch, UInt64 f
...
+    applyDeltas(new ArrayList<>(Collections.nCopies(getTotalTrackedNodeCount(), 0L)));
```
Comment (ajsutton) [inline protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java]: Add comment to explain.

File: protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java [hunk 4]
```diff
@@ -135,18 +138,345 @@ void findHead_shouldAlwaysConsiderJustifiedNodeAsViableHead() {
...
+  @Test
+  void findOptimisticHead_shouldIncludeValidBlocks() {
+    final Bytes32 block1Hash = dataStructureUtil.randomBytes32();
+    final Bytes32 block2Hash = dataStructureUtil.randomBytes32();
+    addBlock(1, block1Hash, GENESIS_CHECKPOINT.getRoot(), VALID);
+    addBlock(2, block2Hash, block1Hash, VALID);
+
+    final ProtoNode head =
+        protoArray.findOptimisticHead(GENESIS_CHECKPOINT.getRoot(), UInt64.ZERO, UInt64.ZERO);
+    assertThat(head).isEqualTo(protoArray.getProtoNode(block2Hash).orElseThrow());
```
Comment (ajsutton) [inline protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java]: Use assertOptimisticHead

File: protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java [hunk 5]
```diff
@@ -135,18 +138,345 @@ void findHead_shouldAlwaysConsiderJustifiedNodeAsViableHead() {
...
+
+  @Test
+  void findOptimisticHead_shouldExcludeBlocksDescendedFromAnInvalidBlock() {
+    final Bytes32 block1Hash = dataStructureUtil.randomBytes32();
+    final Bytes32 block2Hash = dataStructureUtil.randomBytes32();
+    final Bytes32 block3Hash = dataStructureUtil.randomBytes32();
+    addBlock(1, block1Hash, GENESIS_CHECKPOINT.getRoot(), VALID);
+    addBlock(2, block2Hash, block1Hash, OPTIMISTIC);
+    addBlock(3, block3Hash, block2Hash, OPTIMISTIC);
+    protoArray.applyScoreChanges(computeDeltas(), UInt64.ZERO, UInt64.ZERO);
```
Comment (ajsutton) [inline protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java:269]: Add comment to explain this is triggering the update of best descendant index.

File: protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java [hunk 6]
```diff
@@ -135,18 +138,345 @@ void findHead_shouldAlwaysConsiderJustifiedNodeAsViableHead() {
...
+
+    // Check the best descendant has been updated
+    assertThat(protoArray.getProtoNode(block1Hash).orElseThrow().getBestDescendantIndex())
+        .isEqualTo(protoArray.getIndexByRoot(block3Hash));
+    assertOptimisticHead(block3Hash);
+
+    protoArray.markNodeInvalid(block2Hash);
```
Comment (ajsutton) [inline protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java]: Add another test like this where only block3 is marked invalid and block 2 should be the head.

File: protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java [hunk 7]
```diff
@@ -135,18 +138,345 @@ void findHead_shouldAlwaysConsiderJustifiedNodeAsViableHead() {
...
+    voteUpdater.putVote(UInt64.ONE, new VoteTracker(block1bHash, block2aHash, UInt64.ONE));
+    protoArray.applyScoreChanges(computeDeltas(), UInt64.ZERO, UInt64.ZERO);
+
+    // Votes for 2a don't count because it's invalid so we stick with chain b.
+    assertOptimisticHead(block2bHash);
+    assertStrictHead(block1bHash);
+  }
+
+  @Test
+  void updateValidity_shouldRemoveWeightWhenBlocksMarkedAsInvalid() {
```
Comment (ajsutton) [inline protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java]: Didn't wind up calling the method updateValidity

File: protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java [hunk 8]
```diff
@@ -135,18 +138,345 @@ void findHead_shouldAlwaysConsiderJustifiedNodeAsViableHead() {
...
+    // So we switch back to chain b (notably without having to applyScoreChanges)
+    assertOptimisticHead(block2bHash);
+    assertStrictHead(block1bHash);
+  }
+
+  @Test
+  void updateValidity_shouldConsiderAllAncestorsValidWhenMarkedAsValid() {
+    final Bytes32 block1Hash = dataStructureUtil.randomBytes32();
+    final Bytes32 block2Hash = dataStructureUtil.randomBytes32();
+    final Bytes32 block3Hash = dataStructureUtil.randomBytes32();
```
Comment (ajsutton) [inline protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java]: Could probably reuse these.

**Review by rolfyone — APPROVED** [review]
LGTM after the pair review comments addressed.

**Pull request merged**